File: mobile/lib/generation/lesson_validator.dart
import '../models/lesson.dart';

class LessonValidationResult {
  const LessonValidationResult({required this.ok, required this.errors});

  final bool ok;
  final List<String> errors;
}

class LessonValidator {
  const LessonValidator();

  LessonValidationResult validate(Lesson lesson) {
    final errors = <String>[];

    if (lesson.schemaVersion.isEmpty) {
      errors.add('Missing lesson.schemaVersion.');
    }
    if (lesson.id.isEmpty) {
      errors.add('Missing lesson.id.');
    }
    if (lesson.title.isEmpty) {
      errors.add('Missing lesson.title.');
    }
    if (lesson.summary.isEmpty) {
      errors.add('Missing lesson.summary.');
    }
    if (lesson.steps.isEmpty) {
      errors.add('Lesson must include at least one step.');
    }

    for (final indexed in lesson.steps.indexed) {
      _validateStep(indexed.$2, indexed.$1, lesson.source.frameCount, errors);
    }

    return LessonValidationResult(ok: errors.isEmpty, errors: errors);
  }

  void _validateStep(
    LessonStep step,
    int index,
    int frameCount,
    List<String> errors,
  ) {
    if (step.id.isEmpty) {
      errors.add('steps[$index] missing id.');
    }
    if (step.order < 1) {
      errors.add('steps[$index].order must be a positive integer.');
    }
    if (step.title.isEmpty) {
      errors.add('steps[$index] missing title.');
    }
    if (step.instruction.isEmpty) {
      errors.add('steps[$index] missing instruction.');
    }
    if (step.frameIndex < 0 || step.frameIndex >= frameCount) {
      errors.add('steps[$index].frameIndex is outside the extracted frames.');
    }

    final target = step.action.target;
    if (!isRelativeCoordinate(target.x)) {
      errors.add('steps[$index].action.target.x must be between 0 and 1.');
    }
    if (!isRelativeCoordinate(target.y)) {
      errors.add('steps[$index].action.target.y must be between 0 and 1.');
    }
    final width = target.width;
    if (width != null && !isRelativeCoordinate(width)) {
      errors.add('steps[$index].action.target.width must be between 0 and 1.');
    }
    final height = target.height;
    if (height != null && !isRelativeCoordinate(height)) {
      errors.add('steps[$index].action.target.height must be between 0 and 1.');
    }
  }
}

File: src/server/features/video-to-lesson/lessonValidator.js
import { isRelativeCoordinate, lessonActionTypes } from "../../../shared/lessonSchema.js";

export class LessonValidator {
  validate(lesson) {
    const errors = [];

    if (!lesson || typeof lesson !== "object") {
      return { ok: false, errors: ["Lesson must be an object."] };
    }

    for (const field of ["schemaVersion", "id", "title", "summary", "createdAt", "source"]) {
      if (!lesson[field]) {
        errors.push(`Missing lesson.${field}.`);
      }
    }

    if (!Array.isArray(lesson.steps) || lesson.steps.length === 0) {
      errors.push("Lesson must include at least one step.");
    } else {
      lesson.steps.forEach((step, index) => {
        validateStep(step, index, errors);
      });
    }

    return { ok: errors.length === 0, errors };
  }
}

function validateStep(step, index, errors) {
  if (!step || typeof step !== "object") {
    errors.push(`steps[${index}] must be an object.`);
    return;
  }

  for (const field of ["id", "order", "title", "instruction", "action", "frameIndex"]) {
    if (step[field] === undefined || step[field] === null || step[field] === "") {
      errors.push(`steps[${index}] missing ${field}.`);
    }
  }

  if (!Number.isInteger(step.order) || step.order < 1) {
    errors.push(`steps[${index}].order must be a positive integer.`);
  }

  if (!Number.isInteger(step.frameIndex) || step.frameIndex < 0) {
    errors.push(`steps[${index}].frameIndex must be a non-negative integer.`);
  }

  const action = step.action;
  if (!action || typeof action !== "object") {
    errors.push(`steps[${index}].action must be an object.`);
    return;
  }

  if (!lessonActionTypes.includes(action.type)) {
    errors.push(`steps[${index}].action.type is invalid.`);
  }

  const target = action.target;
  if (!target || typeof target !== "object") {
    errors.push(`steps[${index}].action.target must be an object.`);
    return;
  }

  for (const coordinate of ["x", "y"]) {
    if (!isRelativeCoordinate(target[coordinate])) {
      errors.push(`steps[${index}].action.target.${coordinate} must be between 0 and 1.`);
    }
  }

  for (const dimension of ["width", "height"]) {
    if (target[dimension] !== undefined && !isRelativeCoordinate(target[dimension])) {
      errors.push(`steps[${index}].action.target.${dimension} must be between 0 and 1.`);
    }
  }
}
